File: android/src/main/java/com/batch/batch_flutter/BatchFlutterPlugin.java
package com.batch.batch_flutter;

import android.app.Activity;
import android.app.Application;
import android.content.Context;
import android.content.Intent;

import androidx.annotation.NonNull;
import androidx.annotation.VisibleForTesting;
import androidx.core.content.ContextCompat;

import com.batch.android.Batch;
import com.batch.batch_flutter.interop.BatchBridge;
import com.batch.batch_flutter.interop.BatchBridgeException;
import com.batch.batch_flutter.interop.BatchBridgeNotImplementedException;
import com.batch.batch_flutter.interop.BatchBridgePublicErrorCode;

import java.lang.ref.WeakReference;
import java.util.HashMap;
import java.util.Map;

import io.flutter.embedding.engine.plugins.FlutterPlugin;
import io.flutter.embedding.engine.plugins.activity.ActivityAware;
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding;
import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.MethodChannel.MethodCallHandler;
import io.flutter.plugin.common.MethodChannel.Result;
import io.flutter.plugin.common.PluginRegistry;

/**
 * BatchFlutterPlugin
 */
public class BatchFlutterPlugin implements FlutterPlugin, MethodCallHandler, ActivityAware, PluginRegistry.NewIntentListener {

    private static final String PLUGIN_VERSION_SYSTEM_PROPERTY = "batch.plugin.version";

    private static final String PLUGIN_VERSION = "Flutter/1.4.0";

    private final static BatchPluginConfiguration configuration = new BatchPluginConfiguration();

    private static boolean didCallSetup = false;

    private static boolean manageActivityLifecycle = true;

    /// MethodChannels used to communicate with the plugin
    private MethodChannel coreChannel;
    private MethodChannel userChannel;
    private MethodChannel pushChannel;
    private MethodChannel inboxChannel;
    private MethodChannel messagingChannel;

    /// Current Activity
    @VisibleForTesting
    protected WeakReference<Activity> currentActivity = new WeakReference<>(null);

    static {
        System.setProperty(PLUGIN_VERSION_SYSTEM_PROPERTY, PLUGIN_VERSION);
    }

    @VisibleForTesting
    protected boolean isSetup() {
        return didCallSetup;
    }

    @Override
    public void onAttachedToEngine(@NonNull FlutterPluginBinding flutterPluginBinding) {
        coreChannel = new MethodChannel(flutterPluginBinding.getBinaryMessenger(), "batch_flutter");
        coreChannel.setMethodCallHandler(this);
        userChannel = new MethodChannel(flutterPluginBinding.getBinaryMessenger(), "batch_flutter.user");
        userChannel.setMethodCallHandler(this);
        pushChannel = new MethodChannel(flutterPluginBinding.getBinaryMessenger(), "batch_flutter.push");
        pushChannel.setMethodCallHandler(this);
        inboxChannel = new MethodChannel(flutterPluginBinding.getBinaryMessenger(), "batch_flutter.inbox");
        inboxChannel.setMethodCallHandler(this);
        messagingChannel = new MethodChannel(flutterPluginBinding.getBinaryMessenger(), "batch_flutter.messaging");
        messagingChannel.setMethodCallHandler(this);
    }

    @Override
    public void onDetachedFromEngine(@NonNull FlutterPluginBinding binding) {
        coreChannel.setMethodCallHandler(null);
        userChannel.setMethodCallHandler(null);
        pushChannel.setMethodCallHandler(null);
        inboxChannel.setMethodCallHandler(null);
        messagingChannel.setMethodCallHandler(null);
    }

    //region Method calling

    @Override
    public void onMethodCall(@NonNull MethodCall call, @NonNull Result result) {
        if (!isSetup()) {
            final String message = "batch_flutter's BatchFlutterPlugin.setup() has not been called." +
                    "Please make sure that you followed integration steps, and called this method " +
                    "in your Application subclass' onCreate().\n" +
                    "If you're still encountering an error, check that your application subclass" +
                    "is registered using 'android:name' on <application> in your AndroidManifest.xml.";
            BatchFlutterLogger.e(message);
            result.error(BatchBridgePublicErrorCode.MISSING_SETUP.code,
                    message,
                    null);
            return;
        }

        Activity activity = currentActivity.get();
        if (activity == null) {
            final String message = "batch_flutter isn't attached to an activity.";
            BatchFlutterLogger.e(message);
            result.error(BatchBridgePublicErrorCode.NOT_ATTACHED_TO_ACTIVITY.code,
                    message,
                    null);
            return;
        }

        Map<String, Object> arguments = null;
        if (call.arguments != null) {
            if (isObjectAMapOfStrings(call.arguments)) {
                try {
                    //noinspection unchecked
                    arguments = (Map<String, Object>) call.arguments;
                } catch (ClassCastException ignored) {
                }
            } else {
                final String message = "Bridge message root arguments were not null, but not" +
                        "of Map type. Got: '" + call.arguments.getClass().toString() + "'.";
                BatchFlutterLogger.e(message);
                result.error(BatchBridgePublicErrorCode.BAD_BRIDGE_ARGUMENT_TYPE.code,
                        message,
                        null);
                return;
            }
        }

        if (arguments == null) {
            arguments = new HashMap<>();
        }

        BatchBridge.call(call.method, arguments, activity)
                .setExecutor(ContextCompat.getMainExecutor(activity))
                .then(result::success)
                .catchException(e -> {
                    if (e instanceof BatchBridgeNotImplementedException) {
                        result.notImplemented();
                    } else if (e instanceof BatchBridgeException) {
                        BatchBridgeException bridgeException = (BatchBridgeException) e;
                        result.error(bridgeException.pluginCode.code, bridgeException.description, bridgeException.details);
                    } else {
                        BatchFlutterLogger.e("Unknown bridge error", e);
                        result.error(BatchBridgePublicErrorCode.UNKNOWN_BRIDGE_ERROR.code, "Unknown Batch native bridge error. Please see logcat for more info.", null);
                    }
                });
    }

    //endregion

    //region Activity awareness

    @Override
    public void onAttachedToActivity(@NonNull ActivityPluginBinding binding) {
        attachToActivity(binding);
    }

    @Override
    public void onReattachedToActivityForConfigChanges(@NonNull ActivityPluginBinding binding) {
        attachToActivity(binding);
    }

    @Override
    public void onDetachedFromActivity() {
        detachFromActivity();
    }

    @Override
    public void onDetachedFromActivityForConfigChanges() {
        detachFromActivity();
    }

    @Override
    public boolean onNewIntent(Intent intent) {
        BatchFlutterLogger.d("Activity: onNewIntent");
        Activity activity = currentActivity.get();
        if (activity != null) {
            Batch.onNewIntent(activity, intent);
        }
        return false;
    }

    private void attachToActivity(@NonNull ActivityPluginBinding binding) {
        BatchFlutterLogger.d("Activity: attached. Should manage lifecycle: " + manageActivityLifecycle);
        if (manageActivityLifecycle) {
            binding.addOnNewIntentListener(this);
            Batch.onStart(binding.getActivity());
        }
        currentActivity = new WeakReference<>(binding.getActivity());
    }

    private void detachFromActivity() {
        BatchFlutterLogger.d("Activity: detached. Should manage lifecycle: " + manageActivityLifecycle);
        if (manageActivityLifecycle) {
            Activity activity = currentActivity.get();
            if (activity != null) {
                Batch.onStop(activity);
                Batch.onDestroy(activity);
            }
        }
    }

    //endregion

    //region Public API

    /**
     * Ready the plugin for use.
     * This MUST be called in an {@link android.app.Application} subclass' {@link Application#onCreate()}.
     * <p>
     * Once setup success fully, the configuration cannot be changed anymore
     * as this calls {@link com.batch.android.Batch#start(String)}
     * <p>
     * Note: If setup has never been called, or if Batch wasn't provided an APIKey in the manifest or
     * using {@link BatchFlutterPlugin#getConfiguration(Context)} and {@link BatchPluginConfiguration#setAPIKey(String)},
     * any method call will throw an exception.
     *
     * @return Whether the plugin was successfully setup. Returns true on any subsequent call if one
     * setup call succeeded.
     */
    public static synchronized boolean setup(@NonNull Context context) {
        if (didCallSetup) {
            BatchFlutterLogger.i("BatchFlutterPlugin.setup() has already been called once. Ignoring.");
            return true;
        }

        BatchPluginConfiguration pluginConfiguration = getConfiguration(context);
        if (pluginConfiguration.getApiKey() != null) {
            Batch.start(pluginConfiguration.getApiKey());
            Batch.Messaging.setDoNotDisturbEnabled(pluginConfiguration.getInitialDoNotDisturbState());
            didCallSetup = true;
            return true;
        } else {
            BatchFlutterLogger.e("Could not setup BatchFlutterPlugin: your configuration is " +
                    "invalid. Did you set a non-null APIKey using getConfiguration() or using manifest meta-data?");
            return false;
        }
    }

    /**
     * Get the plugin configuration object.
     * It will be initialized with the previous values, including what has been read from the manifest.
     * <p>
     * Once {@link BatchFlutterPlugin#setup(Context)} has been called, changing values in the returned
     * object will not have any effect.
     */
    public static BatchPluginConfiguration getConfiguration(@NonNull Context context) {
        //noinspection ConstantConditions
        if (context == null) {
            throw new IllegalArgumentException("Cannot call getConfiguration with a null context");
        }

        if (didCallSetup) {
            BatchFlutterLogger.i("BatchFlutterPlugin.setup() has already been called. Any " +
                    "configuration change won't be applied, as they all must be " +
                    "done before calling this method.");
        }

        // Ensure that we read the default values from the manifest before any custom one is set
        // as this can be called before "onAttachedToEngine"
        configuration.initFromManifest(context);
        return configuration;
    }

    /**
     * Set whether BatchFlutterPlugin should automatically manage Batch's activity lifecycle
     * (as in automatically calling {@link Batch#onStart(Activity)} and so on).
     * <p>
     * If you add batch_flutter in a hybrid application (one that mixes native android activities with
     * flutter ones), you should turn this off and register {@link com.batch.android.BatchActivityLifecycleHelper}
     * in your {@link Application} subclass.
     */
    public static void setManageActivityLifecycle(boolean manageActivityLifecycle) {
        BatchFlutterPlugin.manageActivityLifecycle = manageActivityLifecycle;
    }

    //endregion

    private boolean isObjectAMapOfStrings(Object object) {
        if (!(object instanceof Map<?, ?>)) {
            return false;
        }

        Map<?, ?> map = (Map<?, ?>) object;

        for (Object key : map.keySet()) {
            if (!(key instanceof String)) {
                return false;
            }
        }

        return true;
    }
}

File: ios/Classes/BatchFlutterPlugin.swift
import Flutter
import UIKit
import Batch

fileprivate struct Consts {
    static let BridgeVersionEnvironmentVar = "BATCH_BRIDGE_VERSION"
    static let BridgeVersion = "Bridge/1.2"
    
    static let PluginVersionEnvironmentVar = "BATCH_PLUGIN_VERSION"
    static let PluginVersion = "Flutter/1.3.0"
}

/// Batch's Flutter Plugin main class.
/// `BatchFlutterPlugin.setup()` needs be called in
///
/// See `configuration` and `manageBatchLifecycle` for more info on how to configure it.
@objc
public class BatchFlutterPlugin: NSObject, FlutterPlugin {
    
    // MARK: Internal static variables & methods
    
    private static var didCallSetup = false
    
    private static func setupBatchEnvironmentVariables() {
        setenv(Consts.BridgeVersionEnvironmentVar, Consts.BridgeVersion, 1)
        setenv(Consts.PluginVersionEnvironmentVar, Consts.PluginVersion, 1)
    }
    
    public static func register(with registrar: FlutterPluginRegistrar) {
        let instance = BatchFlutterPlugin()
        registerChannel(name: "batch_flutter", registrar: registrar, pluginInstance: instance)
        registerChannel(name: "batch_flutter.user", registrar: registrar, pluginInstance: instance)
        registerChannel(name: "batch_flutter.push", registrar: registrar, pluginInstance: instance)
        registerChannel(name: "batch_flutter.inbox", registrar: registrar, pluginInstance: instance)
        registerChannel(name: "batch_flutter.messaging", registrar: registrar, pluginInstance: instance)
    }
    
    private static func registerChannel(name: String, registrar: FlutterPluginRegistrar, pluginInstance: BatchFlutterPlugin) {
        let channel = FlutterMethodChannel(name: name, binaryMessenger: registrar.messenger())
        registrar.addMethodCallDelegate(pluginInstance, channel: channel)
    }

    // MARK: Internal instance variables & methods
    
    private let bridge = Bridge()
    
    public func handle(_ call: FlutterMethodCall, result flutterResult: @escaping FlutterResult) {
        if !isSetup {
            let errorMessage = """
                batch_flutter's BatchFlutterPlugin.setup() has not been called.
                Please make sure that you followed the integration steps, and called this method
                in your AppDelegate's didFinishLaunchingWithOptions method.
                """
            BatchFlutterLogger.logDebug(module: "Bridge", message: errorMessage)
            flutterResult(FlutterError(code: BridgeError.ErrorCode.missingSetup.rawValue,
                                       message: errorMessage,
                                       details: nil))
            return
        }
        
        // We only support [String: AnyObject] arguments, or nil.
        var bridgeParameters: [String: AnyObject] = [:]
        
        if let callArguments = call.arguments {
            if let dictionaryArguments =  callArguments as? [String: AnyObject] {
                bridgeParameters = dictionaryArguments
            } else {
                let errorMessage = "Bridge message root arguments were not nil, but not [String: AnyObject]."
                BatchFlutterLogger.logDebug(module: "Bridge", message: errorMessage)
                flutterResult(FlutterError(code: BridgeError.ErrorCode.badArgumentType.rawValue,
                                           message: errorMessage,
                                           details: nil))
                return
            }
        }
        
        bridge.call(rawAction: call.method, parameters: bridgeParameters)
            .continueOn(bridgeExecutorQueue)
            .then { bridgeResult in
                BatchFlutterLogger.logDebug(module: "Bridge", message: "Got Batch Flutter Call: \(call.method). Result: \(String(describing: bridgeResult))")
                flutterResult(bridgeResult)
            }
            .catch { error in
                if let internalError = error as? BridgeInternalError {
                    if internalError == BridgeInternalError.notImplemented {
                        flutterResult(FlutterMethodNotImplemented)
                        return
                    } else {
                        let errorMessage = "Internal Batch native bridge error (\(internalError)). Please see the console for more info."
                        BatchFlutterLogger.logDebug(module: "Bridge", message: errorMessage)
                        flutterResult(FlutterError(code: BridgeError.ErrorCode.unknownBridgeError.rawValue,
                                                   message: errorMessage,
                                                   details: nil))
                        return
                    }
                } else if let bridgeError = error as? BridgeError {
                    flutterResult(FlutterError(code: bridgeError.code.rawValue,
                                               message: bridgeError.description,
                                               details: bridgeError.details))
                    return
                }
                flutterResult(FlutterError(code: BridgeError.ErrorCode.unknownBridgeError.rawValue,
                                           message: "Unknown Batch native bridge error (\(String(describing: error))). Please see the console for more info.",
                                           details: nil))
                return
            }
    }
    
    internal var isSetup: Bool {
        return BatchFlutterPlugin.didCallSetup
    }
    
    /// Allows the tests to deice how the bridge executor callbacks will be ran
    /// This allows simple calls to be tested (not ones that actually require threading)
    internal var bridgeExecutorQueue: LightPromiseExecutor = LightPromiseExecutor.asyncMain
    
    /// Starts the SDK is the plugin is in managed mode. Does nothing otherwise.
    internal static func startManagedNativeSDK() {
        if manageBatchLifecycle {
            if let batchAPIKey = configuration.actualAPIKey {
                Batch.start(withAPIKey: batchAPIKey)
                BatchPush.refreshToken()
            } else {
                BatchFlutterLogger.logDebug(module: "Bridge", message: "Attempted to start Batch without an apikey, which we had beforehand")
            }
        }
    }
    
    // MARK: Public API
    
    /**
     Get the plugin configuration object.
     It will be initialized with fallback default values.
     
     Once `setup()` has been called, changing values in the object returned by thris property will not have any effect.
     */
    public private(set) static var configuration: BatchPluginConfiguration = BatchPluginConfiguration()
    
    /**
     Set whether BatchFlutterPlugin should automatically manage Batch's lifecycle, as in automatically start.
     
     If you add batch_flutter in a hybrid application, you should turn this off and start/configure Batch natively.
     */
    public static var manageBatchLifecycle: Bool = true
    
    /**
     Prepare Batch's Flutter Plugin for use.
     This must be called in `application(_:didFinishLaunchingWithOptions:)`.
     Failure to do so will throw exceptions on the Flutter side of the plugin.
     
     Once setup succeeds, the configuration cannot be changed using the `configuration` property anymore.
     
     If `manageBatchLifecycle` is true, this method will start Batch and refresh the push token.
     
     - Returns: True if the plugin was setup, false othwersie. Returns true on any subsequent call if one setup call succeeded.
     */
    @discardableResult
    public static func setup() -> Bool {
        if didCallSetup {
            BatchFlutterLogger.logPublic(module: "Plugin", message: "BatchFlutterPlugin.setup() has already been called. Ignoring extra setup call.");
            return true
        }
        
        if !configuration.apply() {
            BatchFlutterLogger.logDebug(module: "Plugin", message: "Could not setup BatchFlutterPlugin: your configuration is invalid. Did you set a non-null APIKey using the configuration var or using Info.plist keys?")
            return false
        }
        
        didCallSetup = true
        
        startManagedNativeSDK()
        
        setupBatchEnvironmentVariables()
        
        return true
    }
}
